### DISCUSSION TITLE
Xstate icons for VSCode

### DISCUSSION THREAD

**Discussion by mmontoya** [statelyai/xstate#1932] [Show and tell]
In answer to a question as to how I customized my VSCode workspace to associate a 'machines' folder with Xstate icons, here's how I did it:

<img width="335" alt="Screen Shot 2021-02-13 at 8 46 34 PM" src="https://user-images.githubusercontent.com/4959569/107887235-4f6e2600-6eca-11eb-8af3-f11692dfe250.png">

1) I went to the Xstate home page and downloaded the logo, which happens to be in svg format (also used by Material Theme icons). [x-state logo](https://xstate.js.org/logo-white.svg)

2) I opened it in Sketch (but any vector editing program, such as Inkscape or Illustrator will do) to remove the "STATE" portion of the logotype. I also added a transparent square frame to enforce a margin around the icon since VSCode will proportionally scale your logo to the maximum size allowable (in either the width or height and this means that the icon will appear too large relative to other icons in the theme, if you don't do this). I then colored the "X" to suit my preference (I usually code in 'dark mode', so I wanted something that stands out while not being too contrasty, so I went with a shade of blue).

3) I am using Phillip Kief's beautiful and very tastefully done "Material Icon Theme". (Thank you, Phillip!). To add custom icons to this theme, you must add the following override to you 'setting.json' file:
```	
"material-icon-theme.files.associations": {
   "*.machine.js": "../../../icons/x-state",
}
```
This will recognize any file with the pattern *.machine.js' as a state machine and apply our newly created icon in the workspace. The location of this icon is relative to the extension's path, which you will find in your home directory under '.vscode/extensions/'  While you can certainly place the icon inside the 'pkief.material-icon-theme-<version>' folder, I wanted to keep my own modification separate from the extension itself (so that it doesn't get lost with removal/updates) so I created an 'icons' folder alongside the 'extensions' directory.  I then named the new icon 'x-state.svg' and placed it inside this folder (specifying the ".svg" is not required). 

Note: I was looking for a way to automatically match filenames with the *Machine.js pattern, but it seems this type of pattern matching hasn't yet been enabled by the theme creator, so you must add a '.' as a separator in the filename).

4) For the folders, you will need a separate svg icon to represent the open and closed state - for this, I fished one out of the theme's icon folder and reworked it to include the Xstate "X" and recolored it to match the X icon's fill color and placed the two states inside my custom 'icons' folder under '.vscode'. I named these:  'folder-machines.svg'  and 'folder-machines-open.svg'. The override in your settings.json file, if you're observing my paths, should be:
```
"material-icon-theme.folders.associations": {
   "customFolderName": "machines",
    "machines": "../../../../../icons/folder-machines"

	}, 
```
That's all you need, just save your newly modified 'setting.json' file to pick up your changes. VSCode will sign your icon filenames with a suffix, so if you later decide to update your custom icons with new ones, you might need to outright delete them and refresh your settings file to pick up the changes.

If you'd like to use the ones I created, here they are:

[icons.zip](https://github.com/davidkpiano/xstate/files/5978431/icons.zip)

You can find more detailed information on what other modifications are available to the theme, from Phillip, its creator,
 [here](https://github.com/PKief/vscode-material-icon-theme#custom-svg-icons).

Many thanks to David Khourshid for his incredible contribution to the webDev community!

Happy coding!

**Comment by tomByrer** [comment]
Thanks for the file & instructions!

> you must add a '.' as a separator in the filename

I do that by convention anyhow.  I try to stay away from capitals in filenames due to using both Windows & Linux.  (React Component Capitals file names trigger my OCD ;) )

**Comment by OlaoluwaM** [comment]
Great! Thanks
